File: app/src/main/java/com/example/petridelivery/lists/datasources/abs/IntegerKeyPagingSource.java
package com.example.petridelivery.lists.datasources.abs;

import android.view.View;

import androidx.paging.PagingSource;
import androidx.paging.PagingState;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

import lombok.AllArgsConstructor;

@AllArgsConstructor
public abstract class IntegerKeyPagingSource<T> extends PagingSource<Integer, T> {
    protected View v;

    @Nullable
    @Override
    public Integer getRefreshKey(@NotNull PagingState<Integer, T> state) {
        Integer anchorPosition = state.getAnchorPosition();
        if (anchorPosition == null) {
            return null;
        }

        LoadResult.Page<Integer, T> anchorPage = state.closestPageToPosition(anchorPosition);
        if (anchorPage == null) {
            return null;
        }

        Integer prevKey = anchorPage.getPrevKey();
        if (prevKey != null) {
            return prevKey + 1;
        }

        Integer nextKey = anchorPage.getNextKey();
        if (nextKey != null) {
            return nextKey + 1;
        }

        return null;
    }

    protected Integer getCurrentPage(LoadParams<Integer> loadParams){
        Integer page = loadParams.getKey();
        return page != null ? page : 1;
    }

    protected Integer getPreviousPage(Integer currentPage){
        return currentPage > 0 ? currentPage - 1 : null;
    }

    protected Integer getNextPage(Integer currentPage, List<T> response){
        return !response.isEmpty() ? currentPage + 1 : null;
    }
}

File: app/src/main/java/com/example/petridelivery/lists/datasources/abs/IntegerKeyPagingSource.kt
package com.example.petridelivery.lists.datasources.abs


import androidx.paging.PagingSource
import androidx.paging.PagingState

abstract class IntegerKeyPagingSource<T: Any> : PagingSource<Int, T>() {

    override fun getRefreshKey(state: PagingState<Int, T>): Int? {
        val anchorPosition = state.anchorPosition ?: return null
        val (_, prevKey, nextKey) = state.closestPageToPosition(anchorPosition) ?: return null
        if (prevKey != null) {
            return prevKey + 1
        }
        return if (nextKey != null) {
            nextKey + 1
        } else null
    }

    protected fun getCurrentPage(loadParams: LoadParams<Int>): Int {
        val page = loadParams.key
        return page ?: 1
    }

    protected fun getPreviousPage(currentPage: Int): Int? {
        return if (currentPage > 0) currentPage - 1 else null
    }

    protected fun getNextPage(currentPage: Int, response: MutableList<T>): Int? {
        return if (!response.isEmpty()) currentPage + 1 else null
    }
}